### ISSUE TITLE
High memory spike on load and save operations sometimes causes out of memory

### ISSUE THREAD

**Issue by stephen** [automerge/automerge#705]
I was investigating why my app sometimes crashes on iOS and I think it's likely that sometimes I'm running into safari iOS' memory limit (seems like ~1.5GB on my device, but probably lower on lower-end devices). I've observed that for my 670KiB document (~5k changes), the `load()` operation uses ~500MiB of RAM, keeping ~100MiB after the document load is complete. These numbers seem high to me (i.e. takes 1000x the file size in memory).

![image](https://github.com/automerge/automerge/assets/2320890/ff82278d-5f13-4347-a8d1-b0f23f5be759)

`save()` operations also seem pretty expensive: memory peaks to ~700MiB allocated, so frequent saves (i.e. an autosave in the background implementation) can sometimes be enough to run out of memory. 

<img width="1436" alt="image" src="https://github.com/automerge/automerge/assets/2320890/44339d22-9d6d-488f-85c0-7c15075301e6">

This seems likely related to issues brought up in #596 and #644, and perhaps could be solved by not loading the entire doc into memory (https://github.com/automerge/automerge/issues/359)?

**Comment by alexjg** [comment]
I'm currently looking into another instance of this. I suspect my analysis in #596 is incorrect because I have a case where memory usage is high for a document with a relatively small number of objects. If you are able to send me the document in question that would help

**Comment by stephen** [comment]
just sent you a slack dm about sharing a doc!

**Comment by stephen** [comment]
Currently, the main way I encounter this issue is when trying to load an automerge doc on first launch of the page. It's been kind of a pain trying to track down how exactly I can trigger this issue and get more details, since they happen before I can attach a debugger to the "home screen web app" in safari.

<img width="898" alt="image (1)" src="https://github.com/automerge/automerge/assets/2320890/f8472522-a2b6-41d7-aa8d-c968b5c97287">

I see this OOM pretty much exclusively on iPhone/iPad Safari (17.1), but seemingly randomly (maybe 1/50 times opening the pwa?).

**Comment by alexjg** [comment]
Thanks for continuing to post the problems you're encountering here, it's useful. We're busy working on performance. The first part of that is some work @orionz has been doing to consolidate all our allocations into a single place. This will enable other work we can do to efficiently generate changes on demand so we don't need to hold on to them in memory after verifying heads. This work will also help us to get a birds eye view of where most of our memory usage is happening which will make it easier to constrain.

In short, addressing this concern is at the top of our stack and we're making progress :slightly_smiling_face:

**Comment by stephen** [comment]
Just wanted to give an update after the recent changes in 2.1.10 (perhaps #800, #815, #808 related).

Happily, the OOM on load issue mentioned in https://github.com/automerge/automerge/issues/705#issuecomment-1802892917 seems to be gone completely, and overall mobile load times have dropped dramatically (7-8s to ~1s) to match behavior on desktop. 🎉 Perhaps the mobile platform really didn't like all the small allocations?

As an overall update, I rebenchmarked the same doc to see how things are different.

Load memory usage is about the same as before: takes ~500MB (still seems pretty high I suppose) on load for the 671KB doc:

<img width="749" alt="image" src="https://github.com/automerge/automerge/assets/2320890/74e6194b-152f-453e-8426-c23c9b5c20eb">

Save seems happier than before, not taking large allocations:
<img width="750" alt="image" src="https://github.com/automerge/automerge/assets/2320890/0c94efc4-29d8-45e7-9573-267be2905ae5">

Really appreciate the work y'all have been doing on this front.

**Comment by alexjg** [comment]
Glad to hear this is improving. Memory usage is next on our list, expect to see improvements there in January.

**Comment by stephen** [comment]
I've been thinking about breaking up my large monolithic automerge document into separate chunks that the application has to dynamically manage loading/unloading for. As part of that I did some rough experiments to see what the current memory footprint for a given automerge doc size looks like.

I roughly did:
```js
Automerge = require('@automerge/automerge')
docs = []
docs.push(Automerge.load(fs.readFileSync('xxx')))
```
and measured memory footprint after running the load call several times.

Results:
<img width="1048" alt="image" src="https://github.com/automerge/automerge/assets/2320890/18169a82-8f92-4459-aad6-6b4b8f7be0d4">

Roughly, it seems like the steady state memory footprint for a given doc is ~80x the compressed file size. I don't have any other commentary about it, but thought it'd be useful to share.

**Comment by pvh** [comment]
Thanks for measuring this stuff! Keep an eye out: when @orionz's memory use patch reaches a testable state we'd love to have you test it with your data and see how much we've improved.
